### PR TITLE
Add Hail-like AC, AN, and AF computation to variant stats V3

### PR DESCRIPTION
**Pull request by awenocur**
Take ref blocks into account when computing AF, and make two new fields: `info_TILEDB_IAC` and `info_TILEDB_IAN` available to data frames.

### CODE REVIEW

File: libtiledbvcf/src/dataset/tiledbvcfdataset.cc
Comment (gspowley) [inline libtiledbvcf/src/dataset/tiledbvcfdataset.cc:768]: Please capture the `bcf_hdr_append` return code in a separate line and test the return code in the if statement.

File: libtiledbvcf/src/dataset/tiledbvcfdataset.cc
Comment (gspowley) [inline libtiledbvcf/src/dataset/tiledbvcfdataset.cc:1083]: A lot of changes in this file are line wrapping only. Please keep the same line wrapping to reduce noise in the code changes.

File: libtiledbvcf/src/dataset/tiledbvcfdataset.cc
Comment (gspowley) [inline libtiledbvcf/src/dataset/tiledbvcfdataset.cc:777]: Same comment as above.

**Review by gspowley — CHANGES_REQUESTED** [review]
This will take multiple passes to review. Here's the first.

**Review by leipzig — COMMENTED** [review]
i would expect some tests in here that are human readable. right now all I see if `df.shape`.

File: libtiledbvcf/src/read/reader.cc
Comment (gspowley) [inline libtiledbvcf/src/read/reader.cc:186]: Is this `TODO` done?

File: libtiledbvcf/src/stats/variant_stats.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats.cc:896]: This looks overly complicated. Let's review in person.

File: libtiledbvcf/test/inputs/stats/first.vcf
Comment (gspowley) [inline libtiledbvcf/test/inputs/stats/first.vcf:18]: GT value of `0/1/0` looks wrong.

File: libtiledbvcf/src/dataset/tiledbvcfdataset.cc
Comment (gspowley) [inline libtiledbvcf/src/dataset/tiledbvcfdataset.cc:782]: Still seeing line wrapping artifacts.

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:130]: This version of `af` looks a lot like the previous version of `af`. Can they be combined?

Maybe set the default value of `num_samples` to 0 and use that to decide what to return?

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:148]: Good place for `auto`?

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:199]: Same comment as above for combining versions of `af_v3`.

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:330]: missing brackets

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:416]: Is this `TODO` done?

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:429]: Is this `TODO` done?

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.cc:530]: This logic is asking for an intermediate variable.

File: libtiledbvcf/src/stats/variant_stats_reader.h
Comment (gspowley) [inline libtiledbvcf/src/stats/variant_stats_reader.h:61]: Does this comment need to move or be removed?

File: libtiledbvcf/test/inputs/stats/first.vcf
Comment (awenocur) [reply libtiledbvcf/test/inputs/stats/first.vcf:18]: This is a test for a triploid variant.

File: libtiledbvcf/src/stats/variant_stats.cc
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats.cc:897]: after discussion: needs more documentation

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats_reader.cc:130]: There were two variants of `af` already; that exploded to four. Distinguishing between fixed vs. variable samples was already present, and was done this way because the method is small, in a header, and called frequently, whereby branching would have a performance impact.

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats_reader.cc:416]: I'd wait until the following release to ensure that we don't want to reexamine that. From a conversation earlier with @leipzig, he told me to leave it for now.

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats_reader.cc:429]: same as above

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats_reader.cc:530]: This change extends the start to include relevant ref blocks. It does so only for variant stats ≥ 3, and doesn't assign a start position less than the beginning of the contig.

File: libtiledbvcf/src/read/reader.cc
Comment (awenocur) [reply libtiledbvcf/src/read/reader.cc:187]: There's currently a check for the V3 stats array being in use; it might be more efficient for queries against datasets with V3 variant stats, that don't request IAC, IAN, or IAF not to enable sorting; this check isn't performed yet.

File: libtiledbvcf/src/stats/variant_stats_reader.h
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats_reader.h]: It describes the variable `should_tally`

File: libtiledbvcf/src/stats/variant_stats_reader.cc
Comment (awenocur) [reply libtiledbvcf/src/stats/variant_stats_reader.cc:198]: same as above: we wouldn't be making it much less succinct, and would be adding runtime complexity by combining them

**Review by gspowley — APPROVED** [review]
Looks good! 

Let's resolve the offline conversation before merging.

**Pull request merged**
